### ISSUE TITLE
TFEA.ChIP

### ISSUE THREAD

**Issue by LauraPS1** [Bioconductor/Contributions#506]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/LauraPS1/TFEA.ChIP

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
The package needs a `vignettes/` directory with markdown or 'Rnw' (LaTeX) vignette. The github repository needs to contain the source of the package, not the built (e.g., Meta/) components. Compare for instance with a typical Bioconductor package, e.g., `git clone https://git.bioconductor.org/packages/DESeq2`

**Comment by LauraPS1** [comment]
Sorry, it's the first time I write a package and the process is still a bit confusing at times. I've already replaced the content of the repository to that of the source of the package.

**Comment by mtmorgan** [comment]
thanks, the `build` and `inst/doc` directories also need to be removed.

**Comment by LauraPS1** [comment]
OK, I deleted them. Is there anything else that I need to modify?

**Comment by mtmorgan** [comment]
I'll move your package from 'awaiting moderation' to 'under review'. A person will be assigned to look at your package in detail, and will provide a written revue in the next week or so.

**Comment by lshep** [comment]
Thank you for your interest in Bioconductor.  Some general questions before a more thorough review: 

How is this different from the numerous software packages already in Bioconductor. Search `chip` or `chipseq` or `enrich` or `transcription` on the [biocviews page](http://bioconductor.org/packages/release/BiocViews.html#___Software)  to see the numerous packages already available. Based on just description in particularly things like: GSEABase, ChIPXpress, triform, chipenrich,  geecc, enrichmentbrowser, phenotest

There are a number of packages that perform GSEA, I would see if any of these algorithms can be used instead of implementing your own. Utilize the existing methods and functions in current Bioconductor packages unless there is a novel algorithm.  

Glancing at the vignette it seems like there should be at least some integration of GeneSets or GeneSetCollection from the GSEABase package.  And you mention that the output of DESeq2 provides files;  it is therefore not unreasonable to imagine your functions working with and/or extending a DESeq2 DESeqResults object.  In fact, you would probably get a larger scope if you utilized or extended the example in the vignette of DESeq2 showing how to incorporate your functions with their current workflow. 

Please also correct the ERRORs in the build report;  This is not isolated to the example in the man pages, when I run the code in the vignette (eval=FALSE section but we test all code when reviewing) I get the following 
```
When I run the plot_CM(pval_mat_UP)
Error in eval(expr, data, expr_env) : object 'log.adj.pVal' not found

and 

> plot_ES(GSEA.result,LFC = hypoxia.sorted$log2FoldChange,specialTF = TF.hightlight,TF_colors = col)
Error in round(enrichmentTable_other$pval.ES, 3) : 
  non-numeric argument to mathematical function
```
Also don't show the internal code of the plotting function in the vignette; This is just confusing for the user. 

If your algorithms provide new, novel approaches, we will gladly include the package in Bioconductor, but currently there needs to be a stronger integration with existing package and classes. 

Cheers, 
Lori

**Comment by LauraPS1** [comment]
Sorry for the delay, we’ve had a busy week. I’ll address your questions one by one below.

> How is this different from the numerous software packages already in Bioconductor. Search chip or chipseq or enrich or transcription on the biocviews page to see the numerous packages already available. Based on just description in particularly things like: GSEABase, ChIPXpress, triform, chipenrich, geecc, enrichmentbrowser, phenotest

The package TFEA.ChIP is intended to identify transcription factors responsible for the co-regulation of sets of genes identified by differential gene expression analysis of gene expression profiles. In this regard it serves a very different purpose than the indicated packages. In addition, although there are other software packages that aim to identify transcription factor binding sites (TFBS) in sets of co-regulated genes, to our knowledge none of them is based on the use of ChIP-seq derived binding sites. In itself the database of TFBS included in the package TFEA.ChIP is quite unique and not found in other software packages.
Below we provide a brief description of what the suggested packages do in comparison with TFEA.ChIP:

- GSEABase: GSEABase provides classes to support GSEA analysis, but does not perform the analysis itself, which our package does, nor includes gene sets to TFBS enrichment analysis.
- ChIPXpress: ChIPXpress searches gene expression profiles for potential targets of a single given TF (defined by its ChIPseq signal). Thus, in a way this package tries to solve the opposite problem that TFEA.ChIP addresses. Namely, it searches for TF targets in gene expression data using ChIPseq data as input, whereas TFEA.ChIP identifies TFBS searching ChIP seq datasets and using a gene expression profile as input. 
- triform: triform is a peak caller, it works with sequencing reads, and ChIP-Seq dataset inputs. Thus it addresses a completely different problem.
- chipenrich: this package is design to find enriched GO or KEGG terms in ChIP-Seq datasets, while our package looks for transcription factors enriched in a given set of differentially expressed genes, usually from an RNA-Seq experiment. Thus both packages solve different problems.
- geecc: while this package uses the same information as an input (differentially expressed genes) geecc is focused on finding enrichment and correlation between GO terms, while, in contrast, our package focus is on the identification transcription factors enriched in co-regulated genes, a task not addressed by geec.
- enrichmentbrowser: this package is also focused on finding enriched KEEG or GO terms, as the previous case
- phenotest: phenotest links differentially expressed genes (DE) to phenotypes defined by the user. As in the previous two cases, this package aims to identify functional terms enriched in sets of DE genes. In contrast, TFEA.ChIP aims to determine the TF responsible for the co-regulation of the DE genes in gene expression profiles.

In summary, even though several packages use the same input information (as geecc or enrichment browser), they’re not designed to perform the same type of analysis, nor they address the biological question that TFEA.ChIP aims to solve, namely the identification of the TF responsible for the coregulation of DE genes.
TFEA.ChIP uses as input a list of differentially expressed genes from gene profiling experiments and aims to find the transcription factor that are more likely to be responsible for the change between the assayed conditions. To that end, it uses ChIPseq-derived information to perform enrichment analysis on the set of DE genes.
To that purpose, we have already built a TFBS database that links every ChIP-Seq experiment selected to genes they’re bound to. The functions included in TFEA.ChIP to build a database are intended mostly for users that may want to modify or expand that currently existing database with their own ChIP-Seq experiments.
If you’re interested in a more concept oriented explanation, we have a reference manual at your disposal [here](https://github.com/LauraPS1/TFEA-drafts/blob/master/Manual.pdf)

> There are a number of packages that perform GSEA, I would see if any of these algorithms can be used instead of implementing your own. Utilize the existing methods and functions in current Bioconductor packages unless there is a novel algorithm.

TFEA.ChIP is not just another implementation of GSEA to identify enriched functional terms. The goal of our package is the identification of TF responsible for the co-regulation of DE genes. To that end, TFEA.ChIP offers, along with other options, a GSEA-like test to identify significant associations with TFBS datasets derived from ChIP-seq experiments. Thus, as it is the case for most of the packages in bioconductor performing GSEA, TFEA.ChIP is adapted to particular conditions, input information and output results. Whereas most of those packages in bioconductor implementing GSEA are focused on finding enriched GO/KEGG or phenotypic terms from established databases of ontology terms, TFEA.ChIP includes a novel database that integrates TFBS derived from ChIPseq datasets that is not found in any other package nor public database. By using this database it aims to answer a biological question not addressed by any of the current packages.
The only package that seems to be general enough, GSEABase, doesn’t provide an algorithm that actually performs the analysis, only classes and methods to do one, Nor includes the core database that feeds the association tests included in TFEA.ChIP.

> Glancing at the vignette it seems like there should be at least some integration of GeneSets or GeneSetCollection from the GSEABase package. And you mention that the output of DESeq2 provides files; it is therefore not unreasonable to imagine your functions working with and/or extending a DESeq2 DESeqResults object. In fact, you would probably get a larger scope if you utilized or extended the example in the vignette of DESeq2 showing how to incorporate your functions with their current workflow.

The original raw datasets upon which the internal database in TFEA.ChIP is compiled can be easily represented using GeneSet  data structure. This will surely facilitate the dissemination and reutilization of this information in other packages and/or applications. However, the lists of potential TF targets derived from ChIPseq experiments typically include from several hundreds to thousands of genes. Thus, storing all the datasets required by TFEA.ChIP using GeneSet data structures results in a package that exceeds by far the 4MB limit set. In its current form, TFEA.ChIP efficiently stores all the required information in a compact table resulting a large reduction of storage. Moreover, this table allows for fast and efficient retrieval of information during the analysis of data. For this reasons, we believe TFEA.ChIP should keep its current structure. 
On the other hand, to facilitate the access to the raw data we have generated the package [ChIP.GeneSets](https://github.com/LauraPS1/ChIP.GeneSets) that stores our current database employing GeneSet data structure.
Finally regarding the integration with DEseq2, we have added a function to extract the relevant information for our analysis from a DESeqResults object.

> Please also correct the ERRORs in the build report

Already done, the package should work properly now.

> Also don't show the internal code of the plotting function in the vignette; This is just confusing for the user.

There’s no internal code shown in the vignette, the only chunks that could be considered “internal code” are marked as echo=FALSE. The reason to include this code is to give a preview of what the output looks like, since, sadly, HTML objects are to big to include the actual plots that the package generates in the vignette.

I hope you find these answers clarifying, and consider our package's approach useful enough to be included in Bioconductor.

**Comment by lshep** [comment]
Please ensure that the [webhook](https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#adding-a-web-hook) has been initialized and please do a version bump for your package.  You mentioned that there have been fixes but no build report has been generated - if the webhook is initialized then a new build report will be generated on a version bump in the DESCRIPTION file 


I built and checked your package locally: 
Please correct 
* checking R code for possible problems ... NOTE
contingency_matrix: no visible binding for global variable ‘Mat01’
deseq2table: no visible global function definition for ‘data_frame’
get_chip_index: no visible binding for global variable ‘MetaData’
Undefined global functions or variables:
  Mat01 MetaData data_frame

* Checking unit tests...
    * NOTE: Consider adding unit tests. We strongly encourage them. See
      http://bioconductor.org/developers/how-to/unitTesting-guidelines/.

* Checking for T...
    Found T in R/All-functions.R (line 371, column 66)
    * WARNING: Use TRUE instead of T (found in 1 files)



**Vignette:**
#extract vector with names of upregulated genes
#extract vector with names of non-responsive genes
- [ ] Have functions for creating these from the hypoxia object with arguments for thresholds 
- [ ] I still wouldn't include the direct ggplot2 code - I would suggest a snapshot or picture instead of html code 
- [ ] The attempt at progress bar  didn't show until finsihed for GSEA_run. Suggestion in comments about code below


**R code**
- [ ] Please pass through formatR to format code and consistent indenting for start and end of brackets/loops - currently hard to follow chunks of code because of formating 
- [ ] This occurs a lot:
Write seq_len(n) or seq_along(x) rather than 1:n or 1:length(x). This protects against the case when n or length(x) is 0 (which often occurs as an unexpected ‘edge case’ in real code) or negative.
- [ ] Remove all requireNamespace or check if package is loaded - these aren't necessary if properly specified in DESCRIPTION and NAMESPACE 


txt2GR
- [ ] check the arguments at the start of the function 
e.g 
```
stopifnot((is.data.frame(fileMetaData) || is.matrix(fileMetaData) || is.array(MetaData)))
stopifnot(format %in% c("narrowpeak", "macs"))
```
- [ ] Also should explicit check the names of columns if expecting the names to be specified or in particular ordering. 

Doing these checks should also compact the code and make it more readable


GSEA_Shuffling
- [ ] use applys and pre allocate - see sections in [efficient coding](
http://master.bioconductor.org/developers/how-to/efficient-code/) - This should be applied throughout the entire code 


GSEA_run
- [ ] use utils::txtProgressBar()

Please start with these changes  - when you are ready for another review of the package please comment back here with what has been updated or with responses justifying current implementation. Please make sure to do a version bump and get a new build report for updated changes as well. 

Thank you 
Lori

**Comment by LauraPS1** [comment]
Thank you for your suggestions. I’m still a rookie programmer, so your feedback is very helping me a lot to improve the package.

I have corrected all the issues you’ve mention in your previous comment. I’ve also changed the format of the script and added some more comment lines to make it more readable, I hope it’s easier for you to read through the code now.

**Comment by lshep** [comment]
Sorry for the delay.  I will look at your package again soon.

**Comment by lshep** [comment]
Sorry for the delay; I was tied up with tasks related to the recent Bioc release. I locally built and check agains devel R 3.5 and Bioc 3.7 - the  single package builders will be updated within the next few days to automatically be on these versions: 
I get the following ERROR:

```
Test ERROR:
* checking tests ...
  Running ‘runTests.R’
 ERROR
Running the tests in 'tests/runTests.R' failed.
Last 13 lines of output:
  1 Test Suite : 
  TFEA.ChIP RUnit Tests - 11 test functions, 0 errors, 1 failure
  FAILURE in test_preprocessInputData: Error in RUnit::checkEquals(nrow(preprocessInputData(hypoxia)), 17166) : 
    Mean relative difference: 0.003800947
  
  
  Test files with failing tests
  
     test_TFEA.ChIP.R 
       test_preprocessInputData 
  
  
  Error in BiocGenerics:::testPackage(pkgname = "TFEA.ChIP") : 
    unit tests failed for package TFEA.ChIP
  Execution halted

```

Could you please look into this?

**Comment by lshep** [comment]
I am rerunning the build manually to have a report for R 3.5 and Bioc 3.7 that has been updated this morning; I will look at your package again in the next few days.  
Cheers

**Comment by lshep** [comment]
Again I apologize for the delayed response. Please see further comments for revision or clarification. 

**R Code revisions or clarification:**

In general I would check a lot of your nested if/elseif cases to be sure that edge cases and all necessary checks are performed.  

**txt2GR**
- [ ] It seems like if fileMetaData is a matrix or array you covert to a data.frame. Do this conversion upfront which will condense the code as then you don't need multiple if statment and only have to check based on the data.frame formate for appropriate column names and column order. 
- [ ] format == narrowpeak 
You explicitly look for a value of -1 of the first row of data for fileTable; Why is it the case that if these columns didn't exist they would be populated with -1; could they just not be there?  it seems like there should be some format check for fileTable then?  You assume the columns are in the right ordering/format is this standardized by technology output?  I would think it would be possible for a legitamate pvalue or qvalue to be -1. so what is the scenerio fileTable[1, 8] != -1 & fileTable[1, 9] != -1 ? Does narrowpeak have standardized column names that could be checked instead? you give a list of
fields in the man page aren't these standardized so you could check for inclusion in the file? 
- [ ] format == macs 
probably should be checking of the dim of the fileTable rather than the length of accessing the first row dim(fileTable)[1].  Again based on the man page I would expect a file to have 5 fields by design? If this isn't the case why? Seems again that the fileTable should be checked for valid columns/column names?
if it doesnt have 4 or 5 columns is it a bad format? 
- [ ] Can checking be more formalized?


**makeTFBSmatrix**
move the if(geneSetAsInput) inside the for loop so that you don't have to repeat code or convert to one format.  Repeated code is never encouraged and there is only one (or two) lines different from the options

 
**preprocessInputData**
We don't encourage direct access with `$` when possible so something like `inputData@listData$log2FoldChange` could be written as `inputData@listData["log2FoldChange"]`  - fix in all code where possible (not always possible)

**Select_genes**
what about the other two cases where min_LFC and max_LFC are either both NULL or both specified - either check for edge cases or handle case or explain why this is not needed. 

**get_chip_index**
- [ ] Can the user provide their own MetaData? why is it checked if it does not exist? should there be some documentation for this? 
-[ ] have a check for input option if database can only be one of those 4 options stopifnot 



Thank you for your time and effort. Please comment back here with clarification and updates.

**Comment by lshep** [comment]
Sorry about the noise,  I was trying to rerun a different package and accidentally ran yours.  The new ERROR that happened to show is an issue on our end.

**Comment by LauraPS1** [comment]
I've made some updates in the package, I'll address your comments point by point below.

> In general I would check a lot of your nested if/elseif cases to be sure that edge cases and all necessary checks are performed.

Done

> txt2GR
It seems like if fileMetaData is a matrix or array you covert to a data.frame. Do this conversion upfront which will condense the code as then you don't need multiple if statment and only have to check based on the data.frame formate for appropriate column names and column order. 

Done

> format == narrowpeak
You explicitly look for a value of -1 of the first row of data for fileTable; Why is it the case that if these columns didn't exist they would be populated with -1; could they just not be there? it seems like there should be some format check for fileTable then? You assume the columns are in the right ordering/format is this standardized by technology output? I would think it would be possible for a legitamate pvalue or qvalue to be -1. so what is the scenerio fileTable[1, 8] != -1 & fileTable[1, 9] != -1 ? Does narrowpeak have standardized column names that could be checked instead? you give a list of fields in the man page aren't these standardized so you could check for inclusion in the file? 

[narrowPeak](https://genome.ucsc.edu/FAQ/FAQformat.html#format12) it’ a standard format, so yes, the 8th and 9th columns should always be dedicated to p-value and q-value respectively, and be populated with ‘-1’ in case they’re not used. On the other hand, narrowPeak files do not need to include column names, that’s why I check the values directly.

> format == macs
probably should be checking of the dim of the fileTable rather than the length of accessing the first row dim(fileTable)[1]. 
Again based on the man page I would expect a file to have 5 fields by design? If this isn't the case why? Seems again that the fileTable should be checked for valid columns/column names?

Yes, bed files generated with MACS have 5 fields by design, but, as it happens with narrowPeak files, they don’t include column names. 

> if it doesnt have 4 or 5 columns is it a bad format? 

Not per se, bed is a widely used format for ChIP-Seq data, but is extremely flexible. It only has three mandatory fields: chromosome, start and stop. In most cases there’s an additional column with the statistical significance of each peak (usually p-value or q-value), another one for peak names (typically something along the lines of ‘MACS_peak_1’, ‘MACS_peak_2’, etc ) and, occasionally, extra columns like peak length, strand, etc. 
The reason to admit input bed files that don’t have the exact same fields as a MACS output is to allow the users to include ChIP-Seq datasets that might be older or generated with another peak caller, as long as it has the minimum information needed.

> Can checking be more formalized? 

I don't think it can be formalized, unless I restrict the input exclusively to either datasets from Encode or direct outputs of MACS, and even then, since these files don’t include column names I’d still need to check for the values stored in the dataset.

> makeTFBSmatrix
move the if(geneSetAsInput) inside the for loop so that you don't have to repeat code or convert to one format. Repeated code is never encouraged and there is only one (or two) lines different from the options

Done

> preprocessInputData
We don't encourage direct access with $ when possible so something like inputData@listData$log2FoldChange could be written as inputData@listData["log2FoldChange"] - fix in all code where possible (not always possible)

Done

> Select_genes
what about the other two cases where min_LFC and max_LFC are either both NULL or both specified - either check for edge cases or handle case or explain why this is not needed.

* In case no p-value or log(Fold Change) limits are defined, the default limits for p-value are 0 – 0.05, thus, the function returns all genes that are differentially expressed (both significantly upregulated and downregulated)
* In case both min_LFC and max_LFC are not defined but min_pval and max_pval are, the function won’t select genes by their log(Fold Change), only those in the range of p-value selected by the user. 
*In case both min_LFC and max_LFC are defined, the function then selects the genes in the input table that have a log(Fold Change) within the limits.

> get_chip_index
Can the user provide their own MetaData? why is it checked if it does not exist? should there be some documentation for this?

Yes, in case the user has generated their own TFBS database or expanded the existing one with their data, and thus needs to use a different metadata object. Several functions check for the existence of user-defined MetaData and Mat01 in the global environment –that’s what the function set_user_data() does, assign the objects the user specifies to the required variable names– and If they’re not defined, the ones included in the package are loaded.

> [ ] have a check for input option if database can only be one of those 4 options stopifnot 

Since the only purpose of that input variable was to filter ChIP-Seq datasets coming from Encode, I’ve change that input variable for a boolean one. I’ve also added an exception in case no ChIP-Seq dataset follows the conditions established by the user.

Thank you for your time

**Comment by lshep** [comment]
Thank you. Please do a version bump to get a new build report and I will look again.

**Comment by LauraPS1** [comment]
I have made a couple changes in the packages:
  * Adding an else condition in a if/else if case that I missed in the function txt2RG.
  * Expand the database with a few more experiments.
  * Adding the possibility to translate gene IDs from mouse to human  in the function GeneID2entrez.

I hope this last change is not a issue for the reviewing process, but if you preferred it if I don't add any new functionalities until it's done, just let me know and I'll reverse the changes in the translating function.

Thank you for your time.

**Comment by lshep** [comment]
Thank you for the updates. The new functionality is okay to add.   I will look at this again today or tomorrow and send a response.  Cheers

**Comment by lshep** [comment]
Some other minor comments: 

**vignette:**
Analyais of the association.... 
- [ ] typo in section i) translate the gen IDs to Entrez Gene IDs  "In case the user es" 
- [ ] And should this be 2 instead of i? 
- [ ] section 3) 
you say all 1122 datasets were analyzed but you could restrict to subset of databases.  How would you restrict the datatbases?  It seems you would have to know what is provided within the internal database in order to properly subset except for built in encodeFilter option? It might be nice to have a helper function that lists the dataset sets and filters accordingly like the prebuilt encode? It seems the filtering is only for all, encode, or specific factors? 

Gene Set Enrichment analysis...
- [ ] For head()  we recommend using [[""]] rather than '$'
```
head(GSEA.result[["Enrichment.table"]])
head(GSEA.result[["RES"]][["GSM2390642"]])
head(GSEA.result[["indicators"]][["GSM2390642"]])
```
Please check example code in vignette and R code for this as well.

building a TF-gene ... 
- [ ] in your example code encourage the best coding practices mentioned:  
Write seq_len(n) or seq_along(x) rather than 1:n or 1:length(x)
Can the for loop be written as vapply? 

**Code**

txt2GR
- [ ] format == mac
There is no closing else if it didn't meet the first three conditions - In which case fileTable and Stat would not be intiialized - have some sort of catch - I get that it would be a very rare case (and probably in ERROR) but say as such. 

GeneID2entrez 
- [ ] An I understanding correctly that if from.Mouse = TRUE, that it translates into human? That should be in the documentation somewhere as well that this is an option 

General: 
- [ ] check for loops and see if any can be vapply's  (may not apply) 
- [ ] See if any of the sections that involve similar code can be condensed by having some conditionals before operations (may not apply) 
- [ ] Update R version in DESCRIPTION to 3.5


Thanks for your continued effort.  Please respond to the above issues and comment back here. 
Cheers,

**Comment by LauraPS1** [comment]
> section 3)
you say all 1122 datasets were analyzed but you could restrict to subset of databases. How would you restrict the datatbases? It seems you would have to know what is provided within the internal database in order to properly subset except for built in encodeFilter option? It might be nice to have a helper function that lists the dataset sets and filters accordingly like the prebuilt encode? It seems the filtering is only for all, encode, or specific factors? 

As is currently implemented, get_chip_index() is intended mainly to filter the database by transcription factor name (i.e., TP53, IRF1, etc.). We added the option to select exclusively ENCODE ChIP-Seqs because some people might be concerned about GEO datasets not having an established pipeline or a required protocol like ENCODE has.
With regard to listing the datasets available, the user can get details about every ChIP-Seq experiment included -TF tested, cell line / tissue, treatment, etc.- loading the metadata table included in the package (using *data("MetaData", package = "TFEA.ChIP")* ). I’ll include this in the vignette so it’s more clear to any user.

I’ve corrected the rest of the things you’ve mentioned, and substituted the for loops in makeTFBSmatrix, contingency_matrix, plot_RES, and GR2tfbs_db for a sapply/lapply. I’ve also added more information about translating mouse gene IDs to human ones in the vignette and function descriptions.

Thanks for your time.

**Comment by lshep** [comment]
Thank you for your commitment to Bioconductor!  Happy Holidays!

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/LauraPS1.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"TFEA.ChIP\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/TFEA.ChIP

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://docs.google.com/forms/d/e/1FAIpQLSdlTbNjsQJDp0BA480vo4tNufs0ziNyNmexegNZgNieIovbAA/viewform
